lib/Backend/IRBuilderAsmJs.cpp

```diff
@@ -599,19 +586,84 @@ IRBuilderAsmJs::RegIsVar(Js::RegSlot reg)
     return false;
 }
 
-BOOL
+bool
 IRBuilderAsmJs::RegIsTypedVar(Js::RegSlot reg, WAsmJs::Types type)
 {
     return reg >= GetFirstVar(type) && reg < GetLastVar(type);
 }
 
+bool
+IRBuilderAsmJs::RegIsTypedConst(Js::RegSlot reg, WAsmJs::Types type)
+{
+    return reg >= GetFirstConst(type) && reg < GetLastConst(type);
+}
+
+bool
+IRBuilderAsmJs::RegIsTypedTmp(Js::RegSlot reg, WAsmJs::Types type)
+{
+    return reg >= GetFirstTmp(type) && reg < GetLastTmp(type);
+}
+
+bool
+IRBuilderAsmJs::RegIs(Js::RegSlot reg, WAsmJs::Types type)
+{
+    return (
+        RegIsTypedVar(reg, type) ||
+        RegIsTypedConst(reg, type) ||
+        RegIsTypedTmp(reg, type)
+    );
+}
+
+bool
+IRBuilderAsmJs::RegIsJitLoopYield(Js::RegSlot reg)
+{
+    return IsLoopBody() && GetJitLoopBodyData().IsYieldReg(reg);
+}
+
+void
+IRBuilderAsmJs::CheckJitLoopReturn(Js::RegSlot reg, IRType type)
+{
+    if (IsLoopBody())
+    {
+#ifdef ENABLE_WASM
+        if (GetJitLoopBodyData().IsLoopCurRegsInitialized())
+        {
+            Assert(m_func->GetJITFunctionBody()->IsWasmFunction());
+            // In Wasm we use the Return opcode for Yields
+            // Check if that yield is outside of the loop (this also covers return)
+            if (!RegIsJitLoopYield(reg))
+            {
+                WAsmJs::Types wasmType = WAsmJs::FromIRType(type);
+                Assert(wasmType < WAsmJs::LIMIT);
+                uint32 typedReg = GetTypedRegFromRegSlot(reg, wasmType);
+                if (GetJitLoopBodyData().IsRegOutsideOfLoop(typedReg, wasmType))
+                {
+                    // The reg should be either a constant (return register)
+                    // Or a temp, in which case, make sure we haven't mapped that temp to something else
+                    Assert(RegIsConstant(reg) || (RegIsTemp(reg) && (GetMappedTemp(reg) == (SymID)reg || GetMappedTemp(reg) == 0)));
+                    GetJitLoopBodyData().SetRegIsYield(reg);
+                    EnsureLoopBodyAsmJsStoreSlot(reg, type);
+                }
+            }
+        }
+        else
+#endif
+        {
+            // In Asm.js there is no Yields outside of the loop, if we're here
+            // It means we are probably returning a value from the loop
```

Comment:
only probably?

Reply:
Humm you're right, it should be definitively a return, I'll fix the comment